[15.0][MIG+REF] portal_sale_personal_data_only: Migration to 15.0 and split in two

yajo opened the pull request:
Pretty basic standard migration. Includes relevant commits from #1776.

Fix https://github.com/OCA/sale-workflow/issues/2236 by extracting these features into the new module `portal_account_personal_data_only`.

The module now depends on it and is auto-installed if `sale` is also installed.

@moduon MT-1451

Keeping in draft until https://github.com/OCA/account-invoicing/pull/1275 is merged.

chienandalu committed:
[ADD] portal_sale_personal_data_only: New module

chienandalu committed:
[MIG] portal_sale_personal_data_only: Migration to 12.0

marylla committed:
Added translation using Weblate (German)

liweijie0812 committed:
Added translation using Weblate (Chinese (Simplified))

HaraldPanten committed:
[IMP] portal_sale_personal_data_only: black, isort, prettier

HaraldPanten committed:
[MIG]portal_sale_personal_data_only: Migration to 13.0

danimv5 committed:
Added translation using Weblate (Spanish)

kirca committed:
[13.0][FIX] portal_sale_personal_data_only: invalid domain in ir.rule

victoralmau committed:
[IMP] portal_sale_personal_data_only: black, isort, prettier

victoralmau committed:
[MIG] portal_sale_personal_data_only: Migration to 14.0

oca-travis committed:
[UPD] Update portal_sale_personal_data_only.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-14.0/sale-workflow-14.0-portal_sale_personal_data_only
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-14-0/sale-workflow-14-0-portal_sale_personal_data_only/

moaadbourhim committed:
[IMP] portal_sale_personal_data_only: black, isort, prettier

moaadbourhim committed:
[MIG] portal_sale_personal_data_only: Migration to 15.0

yajo committed:
[REF] portal_sale_personal_data_only: outsource invoice features

Fix https://github.com/OCA/sale-workflow/issues/2236 by extracting these features into the new module `portal_account_personal_data_only`.

The module now depends on it and is auto-installed if `sale` is also installed.

@moduon MT-1451

pedrobaeza commented on portal_sale_personal_data_only/__manifest__.py:
```diff
@@ -7,7 +7,8 @@
     "website": "https://github.com/OCA/sale-workflow",
     "license": "AGPL-3",
-    "depends": ["sale"],
+    "depends": ["sale", "portal_account_personal_data_only"],
```
Is the dependency over `portal_account_personal_data_only` strictly needed?

yajo replied:
Yes. Due to the split, these rules had to be split in two, one on each of the modules:

https://github.com/OCA/sale-workflow/blob/53cb00f656baf4ce642d683202d7769e49719d0f/portal_sale_personal_data_only/security/security.xml#L19-L34

Without the dependency, usability would be broken. That's also why this module is marked now as auto install.

pedrobaeza replied:
Yes, but then it's not needed technically if I'm not interpreting it badly. I think the `auto_install` is enough, and now yo can put not only `True` / `False`, but the names of the modules to be installed for this one being auto-installed, and you can include there `portal_account_personal_data_only` instead of a hard dependency.

yajo replied:
I see what you mean. I also thought about that. However I'm not sure that removing that dependency is the best idea.

On one side, this is a migration. By adding the dependency, we avoid the need of migration scripts, as this module will behave just as it was before. This is good IMHO as it avoids surprises.

OTOH, according to [the docs](https://www.odoo.com/documentation/15.0/developer/reference/backend/module.html?highlight=auto_install#manifest), I can't do it as you said:

> **If it is a list, it must contain a subset of the dependencies**. This module will automatically be installed as soon as all the dependencies in the subset are installed. The remaining dependencies will be automatically installed as well. If the list is empty, this module will always be automatically installed regardless of its dependencies and these will be installed as well.

If I use a module in `auto_install` that's not among its dependencies, it's not guaranteed to work AFAICS.

So, if I have to choose among removing the dependency or removing the `auto_install` feature, IMHO `auto_install` is more valuable, as it makes the whole experience more coherent. After all, it makes sense that, if you want to restrict sale orders visibility, you want the same on invoices.

In the worst case, this is how it was working before the split, so this is not a regression. If you want I can add the idea to the roadmap.

pedrobaeza replied:
OK, let's keep it as it is.

pedrobaeza commented on portal_sale_personal_data_only/readme/DESCRIPTION.rst:
```diff
@@ -1,7 +1,7 @@
 should stay private.
 
-This module restricts that behaviour so the portal users only see their own
+This module restricts that behavior so the portal users only see their own
```
I usually prefer British English over American:

https://grammarist.com/spelling/behavior-behaviour/

but I think the main reason to let it as it was is for reducing the diff.

yajo replied:
It was just the code spell checker underlining this on red and I simply changed it. Let me undo the change because it is indeed not so important.

OTOH, AFAIK the base lang for odoo is `en_US`. I've never read a guideline on the subject for OCA, but since we follow Odoo's I always understood that's our standard here too. Does it make sense? We need a standard "behavior" in this subject. 🥁 🙃

pedrobaeza replied:
Yes, this has never been discussed, but indeed the base module is American English. We have similar question with for example colour/color. Not sure if Odoo has gone one way or the other. On the other hand, the British one is supposed to be the "template" (not sure if we can compare it with `Spanish (Spain)`). But as said, this one was already put, and it's on the benefit of the diff, but on homogeneity terms, if we see that better everything in American English, let's change it.

I let it in your hands.

yajo replied:
I opened https://github.com/moduon/odoo-community.org/pull/2 as I'm in favor of standardization. However I'll leave it as it is as I'm also in favor of MVP. 😄

rafaelbn commented:
@fcvalgar please could you review? Thank you!

Shide reviewed (approved):
Functional review LGTM

fcvalgar reviewed (approved):
Functional review LGTM

rafaelbn reviewed (commented):
Please @yajo check test! 😄

yajo committed:
[FIX] sale_stock_invoice_plan: wrong import

Tests were generating this error:

```
2022-10-31 10:47:20,850 256 ERROR odoo odoo.tests.loader: Can not `import sale_stock_invoice_plan`.
Traceback (most recent call last):
  File "/opt/odoo/odoo/tests/loader.py", line 33, in _get_tests_modules
    mod = importlib.import_module('.tests', modpath)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/__w/sale-workflow/sale-workflow/sale_stock_invoice_plan/tests/__init__.py", line 3, in <module>
    from . import test_sale_invoice_plan
  File "/__w/sale-workflow/sale-workflow/sale_stock_invoice_plan/tests/test_sale_invoice_plan.py", line 8, in <module>
    from odoo.addons.sale.tests import common, tagged
ImportError: cannot import name 'tagged' from 'odoo.addons.sale.tests' (/opt/odoo/addons/sale/tests/__init__.py)
```

@moduon MT-1451

yajo committed:
[CI] portal_sale_personal_data_only: add rebel group

This module needs to be tested separately, for some others conflict with its tests.

@moduon MT-1451

yajo commented:
Tests fixed. Please somebody merge? Thanks!

chienandalu reviewed (approved):
Thanks! :slightly_smiling_face: :+1:

pedrobaeza commented:
/ocabot migration portal_sale_personal_data_only
/ocabot merge nobump
